File: C/Medium/2134.c
int max(int first, int second)
{
    if (first >= second)
    {
        return first;
    }
    return second;
}

int minSwaps(int *nums, int numsSize)
{
    // use sliding window
    // count the total number of 1s is the size of the required window
    // iterate through the array to get number of swap for a window at a given
    // index

    // use modulo to fix the wrapping problem formed by the circular array
    int totalOnes = 0;
    for (int i = 0; i < numsSize; i++)
    {
        if (nums[i])
        {
            totalOnes++;
        }
    }

    // all ones or no ones in array - edge case so can return
    if (totalOnes == numsSize || totalOnes == 0)
    {
        return 0;
    }

    int currentOnes = 0;
    int maxOnes;

    // first window
    for (int i = 0; i < totalOnes; i++)
    {
        if (nums[i])
        {
            currentOnes++;
        }
    }
    maxOnes = currentOnes;

    // slide the window
    for (int i = 1; i < numsSize; i++)
    {
        // left index
        if (nums[i - 1])
        {
            currentOnes--;
        }

        // right index
        if (nums[(i + totalOnes - 1) % numsSize])
        {
            currentOnes++;
        }
        maxOnes = max(maxOnes, currentOnes);
    }

    return totalOnes - maxOnes;
}

File: Java/Medium/2134.java
class Solution {
    public int max(int first, int second) {
        if (first >= second) {
            return first;
        }
        return second;
    }

    public int minSwaps(int[] nums) {
        // sliding window technique
        // use modulo to wrap for the circular
        // count total ones first
        int totalOnes = 0;
        for (int i : nums) {
            if (i == 1) {
                totalOnes++;
            }
        }

        // edge case:
        // no 1s or all 1s
        if (totalOnes == nums.length || totalOnes == 0) {
            return 0;
        }

        int maxOnes;
        int currentOnes = 0;
        // initial window
        for (int i = 0; i < totalOnes; i++) {
            if (nums[i] == 1) {
                currentOnes++;
            }
        }

        maxOnes = currentOnes;

        // slide the window
        for (int i = 1; i < nums.length; i++) {
            // left edge
            if (nums[i - 1] == 1) {
                currentOnes--;
            }

            // right edge
            if (nums[(i + totalOnes - 1) % nums.length] == 1) {
                currentOnes++;
            }

            maxOnes = max(maxOnes, currentOnes);
        }

        return totalOnes - maxOnes;
    }
}